pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,462 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bitlpm
+
+// Trie is a non-thread-safe binary trie that indexes arbitrarily long
+// bit-based keys with associated prefixes (that is, "masks") indexed from
+// [most significant bit] ("MSB") to [least significant bit] ("LSB") using
+// the [longest prefix match algorithm].
+//
+// Each method's comments describes the mechanism of how the method
+// works.
+//
+// [most significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Most_significant_bit
+// [least significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Least_significant_bit
+// [longest prefix match algorithm]: https://en.wikipedia.org/wiki/Longest_prefix_match
+type Trie[K, T any] interface {
+	// Update updates the trie with a a prefix, key, and value.
+	//
+	// Note:If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Upsert(prefix uint, key K, value T)
+	// Delete removes a key with the given prefix and returns
+	// false if the key was not found.
+	// Note: If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Delete(prefix uint, key K) bool
+	// Lookup lookups any key that matches to at least
+	// the prefix argument.
+	// Note: If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Lookup(prefix uint, key K) T
+	// Len returns the number of entries in the Trie
+	Len() uint
+	// ForEach iterates over every element of the Trie.
+	ForEach(ForEachFunc[K, T])
+}
+
+// ForEachFunc is an interface for using the ForEach method of
+// all Trie types.
```

Comment:
nit: document whether returning true or false halts iteration :-) I always get this wrong.